compiler/rustc_errors/src/lib.rs

```diff
@@ -930,6 +955,40 @@ impl Handler {
     pub fn delay_as_bug(&self, diagnostic: Diagnostic) {
         self.inner.borrow_mut().delay_as_bug(diagnostic)
     }
+
+    pub fn update_unstable_expectation_id(
+        &self,
+        unstable_to_stable: &FxHashMap<LintExpectationId, LintExpectationId>,
+    ) {
+        let diags = std::mem::take(&mut self.inner.borrow_mut().unstable_expect_diagnostics);
+        if diags.is_empty() {
+            return;
+        }
+
+        let mut inner = self.inner.borrow_mut();
+        for mut diag in diags.into_iter() {
+            if let Some(mut unstable_id) = diag.level.get_expectation_id() {
```

Comment:
If I understand correctly, this should always succeed, since you only push "expect"-level diagnostics to `unstable_expect_diagnostics`. Could you add an `else` branch with an panic to ensure that this branch is always actually taken?